server/coordinator/factory.go

```diff
@@ -149,39 +140,44 @@ func (f *Factory) makeCreatePartitionTableProcedure(ctx context.Context, request
 		return nil, err
 	}
 
-	getNodeShardResult, err := request.Cluster.GetNodeShards(ctx)
-	if err != nil {
-		log.Error("cluster get node shard result")
-		return nil, err
-	}
+	snapshot := request.ClusterMetadata.GetClusterSnapshot()
 
 	nodeNames := make(map[string]int)
-	for _, nodeShard := range getNodeShardResult.NodeShards {
-		nodeNames[nodeShard.ShardNode.NodeName] = 1
+	for _, shardNode := range snapshot.Topology.ClusterView.ShardNodes {
+		nodeNames[shardNode.NodeName] = 1
 	}
 
-	partitionTableNum := procedure.Max(defaultPartitionTableNum, int(float32(len(nodeNames))*request.PartitionTableRatioOfNodes))
-
-	partitionTableShards, err := f.shardPicker.PickShards(ctx, request.Cluster.Name(), partitionTableNum, false)
+	subTableShards, err := f.shardPicker.PickShards(ctx, snapshot, len(request.SourceReq.PartitionTableInfo.SubTableNames), true)
 	if err != nil {
-		return nil, errors.WithMessage(err, "pick partition table shards")
+		return nil, errors.WithMessage(err, "pick sub table shards")
 	}
 
-	subTableShards, err := f.shardPicker.PickShards(ctx, request.Cluster.Name(), len(request.SourceReq.PartitionTableInfo.SubTableNames), true)
-	if err != nil {
-		return nil, errors.WithMessage(err, "pick data table shards")
+	shardNodeWithVersion := make([]metadata.ShardNodeWithVersion, 0, len(subTableShards))
+	for _, subTableShard := range subTableShards {
+		for _, shardView := range snapshot.Topology.ShardViews {
```

Comment:
The `ShardView` seems always be used to find shard view by shard id, is it possible to declare it as a map?

Reply:
Modify it to a map, and initialize the map when it is created.